File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;

public class Robot
{
    private static readonly Random Random = new Random();
    private static readonly HashSet<string> AllRobotNames = new HashSet<string>();

    public Robot() => GenerateNewName();

    public string Name { get; private set; }
    
    public void Reset()
    {
        AllRobotNames.Remove(Name);
        GenerateNewName();
    }

    private void GenerateNewName()
    {
        do
        {
            var candidateName = TwoRandomLetters() + ThreeRandomDigits();
            AssignNameIfUnique(candidateName);
        } while (Name == null);
    }

    private static string ThreeRandomDigits() => "" + RandomDigit() + RandomDigit() + RandomDigit();
    
    private static string RandomDigit() => "" + Random.Next(0, 10);
    
    private static string TwoRandomLetters() => RandomLetter() + RandomLetter();
    
    private static string RandomLetter() => ((char)Random.Next('A', '[')).ToString();
    
    private void AssignNameIfUnique(string candidateName) {
        if (AllRobotNames.Add(candidateName)) {
            Name = candidateName;
        }
    }
}

File: clojure/robot-name/src/robot_name.clj
(ns robot-name)


(defn robot []
  (def thing (format "%S%S%d%d%d" (char (+ (rand 26) 65)) (char (+ (rand 26) 65)) (rand-int 9) (rand-int 9) (rand-int 9)))
  )

(defn robot-name [robot]
  thing
)

(defn reset-name [robot]
  (def thing (format "%S%S%d%d%d" (char (+ (rand 26) 65)) (char (+ (rand 26) 65)) (rand-int 9) (rand-int 9) (rand-int 9)))
)
